src/Datadog.Trace/TagValue.cs

```diff
@@ -0,0 +1,200 @@
+using System;
+
+namespace Datadog.Trace
+{
+    /// <summary>
+    /// Span tag value struct, use directly as a string or double
+    /// </summary>
+    public readonly struct TagValue : IEquatable<TagValue>
+    {
+        private readonly string _stringValue;
+        private readonly double? _doubleValue;
+
+        private TagValue(string value)
+        {
+            _stringValue = value;
+            _doubleValue = default;
+            IsMetrics = false;
+        }
+
+        private TagValue(double? value)
+        {
+            _stringValue = null;
+            _doubleValue = value;
+            IsMetrics = true;
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether if the tag value is null
+        /// </summary>
+        public bool IsNull
+        {
+            get
+            {
+                return IsMetrics ? !_doubleValue.HasValue : _stringValue is null;
+            }
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether if the value is a metric
+        /// </summary>
+        public bool IsMetrics { get; }
+
+        /// <summary>
+        /// Gets the string value
+        /// </summary>
+        public string StringValue
+        {
+            get
+            {
+                return _stringValue;
+            }
+        }
+
+        /// <summary>
+        /// Gets the double value
+        /// </summary>
+        public double DoubleValue
+        {
+            get
+            {
+                return _doubleValue ?? default;
+            }
+        }
+
+        /// <summary>
```

Comment:
I think the cast operator should throw if the underlying type isn't the right one. Fetching a double when the tag type is a string feels like a dangerous mistake

Reply:
yep, you're right, I'll do the type check.

Reply:
Ok, so I've changed the TagValue struct to check the `IsMetrics` property before the casting, and also I've change the api to support Nullable<TagValue>.